core/policy/oracle.go

```diff
@@ -0,0 +1,192 @@
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+package policy
+
+import (
+	"fmt"
+	"net/http"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/mysteriumnetwork/node/logconfig/httptrace"
+	"github.com/mysteriumnetwork/node/market"
+	"github.com/mysteriumnetwork/node/requests"
+	"github.com/pkg/errors"
+	"github.com/rs/zerolog/log"
+)
+
+type policySubscription struct {
+	policy      market.AccessPolicy
+	eTag        string
+	subscribers []*Repository
+}
+
+// Oracle represents async policy fetcher from TrustOracle
+type Oracle struct {
+	client             *requests.HTTPClient
+	fetchURL           string
+	fetchInterval      time.Duration
+	fetchLock          sync.RWMutex
+	fetchSubscriptions []policySubscription
+
+	fetchShutdown     chan struct{}
+	fetchShutdownOnce sync.Once
+}
+
+// NewOracle create instance of policy fetcher
+func NewOracle(client *requests.HTTPClient, policyURL string, interval time.Duration) *Oracle {
+	return &Oracle{
+		client:             client,
+		fetchURL:           policyURL,
+		fetchInterval:      interval,
+		fetchSubscriptions: make([]policySubscription, 0),
+		fetchShutdown:      make(chan struct{}),
+	}
+}
+
+// Start begins fetching policies to to subscribers
+func (pr *Oracle) Start() {
+	go pr.fetchLoop()
+}
+
+// Stop ends fetching policies to subscribers
+func (pr *Oracle) Stop() {
+	pr.fetchShutdownOnce.Do(func() {
+		close(pr.fetchShutdown)
+	})
+}
+
+// Policy converts given value to valid policy rule
+func (pr *Oracle) Policy(policyID string) market.AccessPolicy {
+	policyURL := pr.fetchURL
+	if !strings.HasSuffix(policyURL, "/") {
+		policyURL += "/"
+	}
+	return market.AccessPolicy{
+		ID:     policyID,
+		Source: fmt.Sprintf("%v%v", policyURL, policyID),
+	}
+}
+
+// Policies converts given values to list of valid policies
+func (pr *Oracle) Policies(policyIDs []string) []market.AccessPolicy {
+	policies := make([]market.AccessPolicy, len(policyIDs))
+	for i, policyID := range policyIDs {
+		policies[i] = pr.Policy(policyID)
+	}
+	return policies
+}
+
+// SubscribePolicies adds given policies to repository and syncs changes of it's rules from TrustOracle
+func (pr *Oracle) SubscribePolicies(policies []market.AccessPolicy, repository *Repository) error {
+	pr.fetchLock.Lock()
+	defer pr.fetchLock.Unlock()
+
+	subscriptionsNew := make([]policySubscription, len(pr.fetchSubscriptions))
+	copy(subscriptionsNew, pr.fetchSubscriptions)
+
+	for _, policy := range policies {
+		index, exist := pr.getPolicyIndex(subscriptionsNew, policy)
+		if !exist {
+			index = len(subscriptionsNew)
+			subscriptionsNew = append(subscriptionsNew, policySubscription{
+				policy:      policy,
+				subscribers: []*Repository{repository},
+			})
+		}
+
+		var err error
```

Comment:
You don't need to define var before if it used nowhere: 
```suggestion
		if err := pr.fetchPolicyRules(&subscriptionsNew[index]); err != nil {
```